spark2/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreamingRead.java

```diff
@@ -0,0 +1,388 @@
+    File location = new File(parent, "table");
+    File checkpoint = new File(parent, "checkpoint");
+    Table table = createTable(location.toString());
+
+    IcebergSource source = new IcebergSource();
+    List<Long> snapshotIds = SnapshotUtil.currentAncestors(table);
+    Collections.reverse(snapshotIds);
+
+    // test invalid snapshot id
+    DataSourceOptions options = new DataSourceOptions(ImmutableMap.of(
+        "path", location.toString(),
+        "checkpointLocation", checkpoint.toString(),
+        "starting-snapshot-id", "-1"));
+    AssertHelpers.assertThrows("Test invalid snapshot id",
+        IllegalStateException.class, "The option starting-snapshot-id -1 is not an ancestor",
+        () -> source.createMicroBatchReader(Optional.empty(), checkpoint.toString(), options));
+
+    // test specify snapshot-id
+    DataSourceOptions options1 = new DataSourceOptions(ImmutableMap.of(
+        "path", location.toString(),
+        "checkpointLocation", checkpoint.toString(),
+        "starting-snapshot-id", snapshotIds.get(1).toString(),
+        "max-size-per-batch", "1000"));
+    StreamingReader streamingReader = (StreamingReader) source.createMicroBatchReader(
+        Optional.empty(), checkpoint.toString(), options1);
+
+    streamingReader.setOffsetRange(Optional.empty(), Optional.empty());
+    StreamingOffset start = (StreamingOffset) streamingReader.getStartOffset();
+    Assert.assertEquals(start.snapshotId(), snapshotIds.get(1).longValue());
+    Assert.assertEquals(start.index(), 0);
+    Assert.assertTrue(start.isStartingSnapshotId());
+    Assert.assertFalse(start.isLastIndexOfSnapshot());
+
+    StreamingOffset end = (StreamingOffset) streamingReader.getEndOffset();
+    Assert.assertEquals(end.snapshotId(), snapshotIds.get(1).longValue());
+    Assert.assertEquals(end.index(), 1);
+    Assert.assertTrue(end.isStartingSnapshotId());
+    Assert.assertFalse(end.isLastIndexOfSnapshot());
+
+    streamingReader.setOffsetRange(Optional.of(end), Optional.empty());
+    StreamingOffset end1 = (StreamingOffset) streamingReader.getEndOffset();
+    Assert.assertEquals(end1.snapshotId(), snapshotIds.get(2).longValue());
+    Assert.assertEquals(end1.index(), 1);
+    Assert.assertFalse(end1.isStartingSnapshotId());
+    Assert.assertTrue(end1.isLastIndexOfSnapshot());
+
+    streamingReader.setOffsetRange(Optional.of(end1), Optional.empty());
+    StreamingOffset end2 = (StreamingOffset) streamingReader.getEndOffset();
+    Assert.assertEquals(end2.snapshotId(), snapshotIds.get(3).longValue());
+    Assert.assertEquals(end2.index(), 1);
+    Assert.assertFalse(end2.isStartingSnapshotId());
+    Assert.assertTrue(end2.isLastIndexOfSnapshot());
+
+    streamingReader.setOffsetRange(Optional.of(end2), Optional.empty());
+    StreamingOffset end3 = (StreamingOffset) streamingReader.getEndOffset();
+    Assert.assertEquals(end3, end2);
+  }
+
+  private Table createTable(String location) {
```

Comment:
I think this should be in a `@Before` or `@BeforeClass` method. Also, no need for real files for the `getChanges` tests.